operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/KeycloakStatus.java

```diff
@@ -87,6 +88,14 @@ public void setObservedGeneration(Long generation) {
         this.observedGeneration = generation;
     }
 
+    public String getHostname() {
+        return hostname;
+    }
+
+    public void setHostname(String hostname) {
+        this.hostname = hostname;
+    }
```

Comment:
I think exposing hostname in the status might be a bit confusing in the current state. IMHO it's not what was originally desired in #21741 (CC @pgodowski), and I don't see much benefit in exposing it just through status either. I know it's aligned e.g. with how Route resource behaves but I'm not sure it makes so much sense for Keycloak CR if Keycloak actually doesn't use this hostname.

Reading #21741 and also considering how old Operator behaved, I think the main expectation from this feature is that the `hostname` will get a default auto-generated value if not specified by user. I.e. the Operator would set the field on the CR spec directly. Alternatively, pass the auto-generated hostname to Keycloak – i.e. behave as it was set in the CR spec.

On the other hand, hostname could be also baked in the image. The Operator currently has no idea what configuration comes from the CR and what is already set by a different config source, typically a properties file in a custom image. If we specify the auto-generated value in the CR, it might override what is set e.g. in a properties file in custom image.

I can see a few option here but I'm not too happy about either:
1. Check if Keycloak truly has no hostname set (related to #21816). Only if really no hostname is specified by user, Operator generates one.
2. We introduce a new switch in the CR, something like `hostname.generate: true`, that'd make the Operator set a default hostname only if explicitly enabled to do so.
3. We just go ahead and set the default auto-generated hostname when it's not specified in the CR regardless if it's defined in some other config source. In the end, we do something similar e.g. for ports – we also assume only CR values and not other config sources.

I'm probably most inclined towards option 3 to keep it simple. It'd be of course a breaking change which we'd need to document. WDYT?

Reply:
> IMHO it's not what was originally desired in https://github.com/keycloak/keycloak/issues/21741 (CC @pgodowski), and I don't see much benefit in exposing it just through status either.

If the operator takes responsibility for creating the hostname, it should provide that to the user rather than requiring a secondary lookup via the route.

> We introduce a new switch in the CR, something like hostname.generate: true, that'd make the Operator set a default hostname only if explicitly enabled to do so.

Rather than introduce another flag, and especially if this feature is going to be static - that is generated only once and stored on the cr, I'd argue instead for dropping the feature entirely and just make the user responsible for filling in some hostname using the apps domain.

As an aside the fundemental issue here is that openshift is basically violating the contract of ingress.  The ingress docs make it clear that the hostname is optional: https://kubernetes.io/docs/concepts/services-networking/ingress/#ingress-rules - obviously though there would quickly be routing conflicts if that were honored, especially if everything is using the default ingresscontroller.

Reply:
> > IMHO it's not what was originally desired in #21741 (CC @pgodowski), and I don't see much benefit in exposing it just through status either.
> 
> If the operator takes responsibility for creating the hostname, it should provide that to the user rather than requiring a secondary lookup via the route.
> 

While I haven't asked for `.status.hostname` (since the connection binding information would be handled differently in my specific use case), I agree it makes sense to put it back into Keycloak status, instead of the secondary lookup. I understand that this hostname is not only used by Route, but also used by Keycloak operand, thus I think it deserves to be put into Keycloak `.status`.

Reply:
IMHO if the Operator generates a hostname, it should also configure Keycloak to use it, not just expose it in status. We can keep it in status but we should also pass it to the Keycloak instance. If we keep it just in the status without Keycloak actually using it, it might be confusing to the user and would still require them to manually set the hostname in the CR spec.

Now the question is how to do that. For now I think we can assume the CR is the source of truth. I.e. let's omit the use case that user baked in their hostname inside the custom image for this discussion's sake. And I think we both agree that additional flag is not the way to go. Users can always disable the default Ingress completely to avoid this behaviour.

I see now two options.
1. Make this static, i.e. set the generated hostname on the spec. We would basically handle that as a default value that we persist. The Operator could then treat it in a standard way as if it was set by user. We might be even able to not watch the Ingress config for changes in the apps domain, if we are ok with fetching the config every time there's a new CR  without hostname. I think this might be the simplest approach.
2. Make this dynamic and more implicit. Don't touch the spec but still pass the generated hostname to Keycloak instance. We'd probably need to watch the Ingress config for changes in the apps domain in this case.

Reply:
> I understand that this hostname is not only used by Route, but also used by Keycloak operand

That's not the case yet with the current PR. We only expose it in status without configuring the operand to use it.

Reply:
> IMHO if the Operator generates a hostname, it should also configure Keycloak to use it, not just expose it in status.

The generated hostname in this pr is provided to Keycloak, so this does provide what you are calling option 2.

> I think this might be the simplest approach.

I see that your preferred solution is to look up the appsDomain each time and to populate on the CR.  It is simpler, but not quite as simple as you'd think - https://github.com/keycloak/keycloak/pull/22859#discussion_r1327286266 

By doing it this way I think you are also trying to side step having to do a status update.  

To be honest I'm not a big fan of the controllers making spec updates as it can get confusing on round-tripping and makes the stored version of the resource read less intentionally than what the user supplied.

However double checking the openshift logic - there's no expectation that existing routes will change if the appsDomain is changed, but newly created routes will have the updated appsDomain.  The user is expected to update existing ones. 

To corollary here is that if we're not watching the ingress config, then if we populate the hostname someone will have to remove it, or set it to the new expected value.  If we don't populate the hostname and aren't watching, then something will need to trigger another reconciliation.